[Docs] Add Rollback instructions in the Upgrade docs (#21334)
yugabyte/yugabyte-db: docs/content/preview/manage/upgrade-deployment.md

Reviewer: Typo: I think you meant unusual behavior in a few places, but text says usual behavior. Even with the typo fixed, I think unusual sounds odd. How about - “If you encounter any issues after the upgrade, you have the option of Rolling back to the previous version”.

Reviewer: Rest of the document looks fine. Find all occurrences of “Unusual” or “Usual” due to typo and change.

New YugabyteDB features may require changes to the format of data that is sent over the wire or stored on disk. During the upgrade process, these new features are disabled to prevent sending the new data formats to processes that are still running the old version. After all YugabyteDB processes have been upgraded, these features can be safely enabled. [AutoFlags](https://github.com/yugabyte/yugabyte-db/blob/master/architecture/design/auto_flags.md) helps simplify this procedure so that you don't have to worry about identifying what these features are, and how to enable them. All new features are enabled using a single command.

1. Use the [yb-admin](../../admin/yb-admin/) utility to promote the new AutoFlags:

```sh
./bin/yb-admin \
    -master_addresses <master-addresses> \
    promote_auto_flags
```

Expect to see the following output:

```output
PromoteAutoFlags completed successfully
New AutoFlags were promoted
New config version: 2
``````
OR
```output
PromoteAutoFlags completed successfully
No new AutoFlags eligible to promote
Current config version: 1
``````

{{< note title="Note" >}}
- `promote_auto_flags` is idempotent and can be run multiple times.

> - Before promoting AutoFlags, ensure that all YugabyteDB processes in the cluster have been upgraded to the new version. If any process running an old version may fail to connect to the cluster after the AutoFlags have been promoted.

Reviewer 2: ```suggestion
- Before promoting AutoFlags, ensure that all YugabyteDB processes in the cluster have been upgraded to the new version. Any process running an old version may fail to connect to the cluster after the AutoFlags have been promoted.
```

> - Before promoting AutoFlags, ensure that all YugabyteDB processes in the cluster have been upgraded to the new version. If any process running an old version may fail to connect to the cluster after the AutoFlags have been promoted.

Reviewer 2: "YugabyteDB processes" - you mean masters and tservers? or something else?

{{< /note >}}

2. Wait at least 10 seconds (`FLAGS_auto_flags_apply_delay_ms`) for the new AutoFlags to be propagated and applied on all YugabyteDB processes.

#### 2. Upgrade the YSQL system catalog

> If you do not have [YSQL enabled](../../reference/configuration/yb-tserver/#ysql-flags), you can skip this step.

Reviewer 2: ```suggestion
If you do not have [YSQL enabled](../../reference/configuration/yb-tserver/#ysql), you can skip this step.
```

YSQL successfully upgraded to the latest version
```

> In certain scenarios, a YSQL upgrade can take longer than 60 seconds, which is the default timeout value for `yb-admin`. If this happens, run the command with a greater `-timeout_ms` value:

Reviewer 2: ```suggestion
In certain scenarios, a YSQL upgrade can take longer than 60 seconds, which is the default timeout value for `yb-admin`. If this happens, run the command with a greater `-timeout_ms` value. For example:
```

> **Example**

Reviewer 2: ```suggestion
```

```sh
./bin/yb-admin \
    -master_addresses ip1:7100,ip2:7100,ip3:7100 \
    -timeout_ms 180000 \
    upgrade_ysql
```

{{< note title="Note" >}}
- `upgrade_ysql` is idempotent and can be run multiple times without any side effects.
- Concurrent operations in a cluster can lead to transactional conflicts, catalog version mismatches, and read restart errors. This is expected, and should be addressed by retrying the operation. If `upgrade_ysql` encounters these errors, then it should also be retried.
{{< /note >}}

### B. Rollback Phase

{{< note title="Note" >}}

> - You cannot rollback after finalizing the upgrade.

Reviewer 2: ```suggestion
- You cannot perform a rollback after finalizing the upgrade.
```

> - Rollback is only supported when you are upgrading a cluster that is already on version `2.20.2.0` or higher.

Reviewer 2: ```suggestion
- Rollback is only supported when you are upgrading a cluster that is on v2.20.2.0 or later.
```

{{< /note >}}

> In order to rollback to the version that you were on before the upgrade, you need to restart all YB-Master and YB-Tservers on the old version.

Reviewer 2: ```suggestion
To roll back to the version that you were on before the upgrade, you need to restart all YB-Master and YB-TServers on the old version.
```

> #### 1. Rollback YB-Tservers

Reviewer 2: ```suggestion
#### 1. Roll back YB-TServers
```

> Rollback the YB-Tservers one node at a time:

Reviewer 2: ```suggestion
Roll back the YB-TServers one node at a time:
```

1. Stop the `yb-tserver` process.

    ```sh
    pkill yb-tserver
    ```

1. Verify that you're in the directory of the old version.

    ```sh
    cd /home/yugabyte/softwareyb-$OLD_VER/
    ```

1. Start the old version of the YB-TServer process. Follow the instructions from [Start YB-TServers](../../deploy/manual-deployment/start-tservers/).

> 1. Make sure that all YB-Tserver processes are running and the cluster load is balanced at `http://<any-yb-master>:7000/tablet-servers`.

Reviewer 2: ```suggestion
1. Make sure that all YB-TServer processes are running and the cluster load is balanced at `http://<any-yb-master>:7000/tablet-servers`.
```

1. Pause for at least 60 seconds before rolling back the next YB-TServer.

> #### 2. Rollback YB-Masters

Reviewer 2: ```suggestion
#### 2. Roll back YB-Masters
```

> Use the following procedure to rollback all YB-Masters:

Reviewer 2: ```suggestion
Use the following procedure to roll back all YB-Masters:
```

1. Stop the `yb-master` process.

    ```sh
    pkill yb-master
    ```

1. Verify that you are in the directory of the old version.

    ```sh
    cd /home/yugabyte/softwareyb-$OLD_VER/
    ```

1. Start the old version of the YB-Master process. Follow the instructions from [Start YB-Masters](../../deploy/manual-deployment/start-masters/).

> 1. Make sure that all YB-Master processes are running at http://<any-yb-master>:7000/.

Reviewer 2: ```suggestion
1. Make sure that all YB-Master processes are running at `http://<any-yb-master>:7000/`.
```

1. Pause for at least 60 seconds before rolling back the next YB-Master.

## Upgrades with xCluster

When you have unidirectional xCluster replication, it is recommended to upgrade the target cluster before the source. Once the target cluster is upgraded and finalized, you can proceed to upgrade the source cluster.

> If you have bidirectional xCluster replication, then you should upgrade and finalize both clusters at the same time. Run the [Upgrade Phase](#upgrade-phase) for each cluster individually and monitor both of them. If you encounter any unusual behavior, rollback both clusters. If everything looks healthy, finalize both with as little delay as possible.

Reviewer 2: ```suggestion
If you have bidirectional xCluster replication, then you should upgrade and finalize both clusters at the same time. Run the [Upgrade Phase](#upgrade-phase) for each cluster individually and monitor both of them. If you encounter any issues, roll back both clusters. If everything looks healthy, finalize both with as little delay as possible.
```

{{< note title="Note" >}}

> xCluster replication requires the target cluster version to be equal to or higher than the source cluster version. The setup of a new xCluster replication will fail if this check fails. Existing replications will automatically pause if the source is finalized before the target.

Reviewer 2: ```suggestion
xCluster replication requires the target cluster version to be the same or later than the source cluster version. The setup of a new xCluster replication will fail if this check fails. Existing replications will automatically pause if the source is finalized before the target.
```

```

{{< note title="Note" >}}
- `upgrade_ysql` is idempotent and can be run multiple times.
- Concurrent YSQL operations in a cluster can lead to transactional conflicts, catalog version mismatches, and read restart errors. This is expected, and should be addressed by retrying the operation. If `upgrade_ysql` encounters these errors, then it should also be retried.
{{< /note >}}

### B. Rollback Phase

{{< warning title="Important" >}}
- Rollback is only supported when you are upgrading a cluster that is already on version `2.20.2.0` or higher.
- You cannot rollback after finalizing the upgrade. If you still want to go back to the old version, you can [Export and import](../backup-restore/export-import-data/) the data to another cluster that is running the old version. You may have to manually modify the SQL script to make it match the old version's format.
{{< /warning >}}

> In order to rollback to the version that you were on before the upgrade, you need to restart all YB-Master and YB-Tservers on the old version. All YB-Tservers have to be rolled back before you rollback YB-Masters.

Reviewer 2: ```suggestion
In order to roll back to the version that you were on before the upgrade, you need to restart all YB-Master and YB-TServers on the old version. All YB-TServers have to be rolled back before you roll back YB-Masters.
```
